_posts/2024-01-22-Python-Memory-Management-and-Troubleshooting.md

```diff
@@ -0,0 +1,568 @@
+---
+title: "Tip of the Month: Python Memory Management and Troubleshooting"
+author: dave-bunten
+tags:
+  - tip-of-the-month
+  - python
+  - memory
+  - memory-management
+  - memory-allocators
+---
+
+# Tip of the Week: Python Memory Management and Troubleshooting
+
+{% include tip-of-the-month-intro.html %}
+
+## Introduction
+
+<!-- excerpt start -->
+Have you ever run Python code only to find it taking _forever_ to complete or sometime abruptly ending with an error like: `123456 Killed` or `killed (program exited with code: 137)`?
+You may have experienced memory resource or management challenges associated with these scenarios.
+This post will cover some computer memory definitions, how Python makes use of computer memory, and share some tools which may help with these types of challenges.
+<!-- excerpt end -->
+
+## What is Memory?
+
+### Computer Memory
+
+<!-- set a max width for mermaid diagram below so it doesn't render so large -->
+<style>
+.smaller-mermaid-diagram img { max-width: 500px; }
+</style>
+
+<div class="smaller-mermaid-diagram">
+
+```mermaid!
+flowchart LR
+
+subgraph computer ["Computer (resources)"]
+  memory["Memory"]
+  storage["Data Storage"]
+  cpu["CPU(s)"]
+end
+
+style computer fill:#fff,stroke:#333
+style storage fill:#fff,stroke:#333
+style cpu fill:#fff,stroke:#333
+style memory fill:#86EFAC,stroke:#333
+
+```
+
+</div>
+
+_Computer memory is a type of computer resource available for use by software on a computer_
+
+Computer memory, also sometimes known as "RAM" or "random-access memory", or "dynamic memory" is a type of resource used by computer software on a computer.
+"Computer memory stores information, such as data and programs for immediate use in the computer. ... Main memory operates at a high speed compared to non-memory storage which is slower but less expensive and oftentimes higher in capacity. " ([Wikipedia: Computer memory](https://en.wikipedia.org/wiki/Computer_memory)).
+
+<!-- table HTML retained for replicability 
```

Comment:
Should ideally just leave this table as HTML instead of a figure. Better accessibility and scaling/clarity. [Tables can have captions](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/caption) too if that's what you were concerned about, though I don't think I have any nice styles in the template for a table caption. I can help you patch some in here.

Reply:
Thanks for this comment! The main challenge with these tables was that something went wrong when rendering tables within tables. The HTML was otherwise valid so far as I could see, but when rendered through this project there seemed to be wrapped table blocks and divs added which broke the display and tables (corrupting later HTML in the process). Brainstorming: another way to use the HTML without rendered modifications would be to wrap in an iframe somehow, but I felt the images might be less prone to challenges while balancing the desire to publish before the next newsletter (next week).

Does Netlify allow one to view previous deploys? If so, the original commit / deploy version when the PR was opened is probably the best point of reference to see the originally breaking HTML due to the tables within tables.

Would you recommend opening an issue with Lab-website-template? I wasn't sure whether I did something out of bounds for that project, Jekyll Spaceship, or HTML more generally.

Reply:
I'll check out that commit. Feel free to ignore this comment for now to publish before the newsletter.

Reply:
Thanks - just in case it helps with the debug I added a test / debug PR from an earlier point with #29 . See this link for a preview https://cu-dbmi.github.io/set-website/preview/pr-29/2024/01/22/Python-Memory-Management-and-Troubleshooting.html (note the raw HTML content in the first table, and the rendering near the end of the post content which gets jumbled).